apps/email/js/cards/cmp/autocomplete.js

```diff
@@ -0,0 +1,476 @@
+    if (startString) {
+      node.appendChild(document.createTextNode(startString));
+    }
+
+    var matchString = text.substring(index, endIndex);
+    if (matchString) {
+      var span = document.createElement('span');
+      span.classList.add('highlight');
+      span.textContent = matchString;
+      node.appendChild(span);
+    }
+
+    var endString = text.substring(endIndex);
+    if (endString) {
+      node.appendChild(document.createTextNode(endString));
+    }
+  } else {
+    node.textContent = text;
+  }
+}
+
+/**
+ * Autocomplete custom element that contains a list of elements that show the
+ * autocomplete matches.
+ *
+ * This custom element relies on a CSS class-based API to relate nodes outside
+ * of this autocomplete. The CSS/HTML structure expected by this autocomplete:
+ *
+ * - A parentNode that contains both this autocomplete element and the input
+ * elements that want autocomplete.
+ *
+ * - The parentNode must contain an element that holds all the input fields,
+ * with the CSS class of "cmp-autocomplete-input-list". That inputArea will be
+ * sized to know how tall to make the spacer to allow proper scrolling of the
+ * parentNode so that the target input element is at the top of the view.
+ *
+ * - A "cmp-autocomplete-origin" element that is a parent of parentNode that
+ * is the master scroll area for the view. This element is used as the "origin"
+ * for the style.top placement of the autocomplete list and for proper scrolling
+ * of the parentNode.
+ *
+ * The autocomplete custom element also expects a "data-source" HTML attribute
+ * to be set on it, which indicates the module ID to use as the data source for
+ * the matches.
+ *
+ * The owner of this custom element instance (typically a view card), should
+ * override the getExistingEntries() method on the instance to return the
+ * current entries for a given input, to avoid dupes showing up in subsequent
+ * autocomplete matches.
+ *
+ * The owner can also use `instance.topMatch` to get the currently displayed
+ * top match in the autocomplete list.
+ */
+return [
+  require('../base')(),
+  {
+    createdCallback: function() {
+      //TODO: master hack only. In email-model-instance branch, this is not
+      //needed. In the branch, there is a separate 'base' that elements can
+      //mix in that is not assumed to be a card. For now, this is an equivalent
+      //result.
+      this.classList.remove('card');
+
+      this._matchClickedTime = 0;
+
+      this.runQuery = this.runQuery.bind(this);
+      this.onFocus = this.onFocus.bind(this);
+      this.onBlur = this.onBlur.bind(this);
+      this.onInput = this.onInput.bind(this);
+
+      // Bind to events triggered by input field children. Need to explicitly
+      // bind to the elements instead of bubble since focus/blur do not
+      // bubble.
+      var nodes = this.parentNode.querySelectorAll('input');
+      slice.call(nodes).forEach((inputNode) => {
```

Comment:
fyi-you-are-probably-aware-of: Array.from(nodes) is the new hotness over this dance (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from) that we can safely use and is backport safe through gecko 32 but of course is potentially disaster if we don't assume modern-ish Firefox/Chrome.